### ISSUE TITLE
Refused to evaluate a string as JavaScript because 'unsafe-eval'

### ISSUE THREAD

**Issue by devangbhuva97** [strukturag/libheif#173]
I'm getting error - _Refused to evaluate a string as JavaScript because 'unsafe-eval'_. I went through [libheif.js](https://strukturag.github.io/libheif/libheif.js) code and I see, use of **`new Function()`**.

**Note:** I have added Content Security Policy & didn't add `unsafe-eval`.

**Comment by levibuzolic** [comment]
Will need to compile to WASM instead of pure JS using emscripten to get this working. Currently on the same mission.

**Comment by BSteffaniakSS** [comment]
Hi, we are running into the same CSP issue. Has there been any progress on compiling to WASM? Thanks.

**Comment by jamsinclair** [comment]
I _believe_ you can set a flag with emscripten to avoid this, `-s DYNAMIC_EXECUTION=0`. 

Hopefully will give this a try soon and get back to you 😸 🤞 . Could be wrong 😅 .

**Comment by thomas5280** [comment]
Hi everyone. Thanks for this library. Now in 2024, I wanted to share that I Really think we need a build of this that avoids the use of eval(), which requires unsafe-eval applied to the Content Security Policy (CSP).

The issue has gotten clearer over the years. Unsafe-eval is especially problematic, and restricting 'eval' isn't really considered a "high security context" but rather a standard best practice for generally secure websites, right at the same tier as avoid unsafe-inline for scripting.

What's especially challenging is that unsafe-eval cannot be applied on a per-script, or even per-domain basis. Once opened up, it's opened up for everything.

Recently, a WASM-based approach using wasm-unsafe eval for Chrome and other supported browsers has been introduced. But this is not sufficient, because it still permits the use of eval in WASM, which itself increases security risk.
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src#unsafe_webassembly_execution

My own CSP logs, from 'report-sample', indicate a significant number of blocked attempts at the use of wasm-unsafe-eval, including by Chrome Extensions.

In summary, the most secure way I can deliver libhef's functionality to my reasonably secure application is through the use of standard JavaScript compiled library without the use of WASM. While not necessarily the highest performance, it is legitamately more secure in that it does not require the permitting of unsafe-eval in either JS or WASM contexts.

Can you please provide a fourth option of the library for 1.17.2, for example, compiled without the need for eval(),

-s DYNAMIC_EXECUTION=0


...as described here:
https://github.com/emscripten-core/emscripten/blob/1bc49003b9a5310362d2e4a6334a62be9cd56dc2/src/settings.js#L1282

Other than that, can you please provide instructions for how to achieve the same?


Thank you!

**Comment by farindk** [comment]
Could anyone with experience in that area please prepare a PR for this? (cc: @fancycode)

**Comment by thomas5280** [comment]
Hi there -- this continues to plague me in production, and I don't think there is any other solution that is considered fully secure, other than to compile with DYNAMIC_EXECUTION=0.

Is it feasible for end users of this library to compile an emscriptem release themselves? Sometimes this is much more involved than a simple command line. I wasn't able to find docs on this.

Either way -- I think it's an important addition to the community to be able to provide this functionality to end users but without forcing loosening of content security policies.

Can you compile a release with DYNAMIC_EXECUTION=0?

**Comment by thomas5280** [comment]
@farindk Would need your consult time here, request is simple --

1. Create a one-time Emscriptem pure JS single-library HEIF-only release of this; single file noplugins; Emscripten release wth:

a. DYNAMIC_EXECUTION=0 and USE_WASM=0 (don't allow unsafe-eval so it works with CSPs that restrict unsafe-eval). 

b. Output should be a single JS file, not Node-JS, to be embedded in a standard HTML file like the demo app to support plain vanilla javsascript application. Single JS include in <script> tags, no other plugs, self contained, no dynamic loading, no require-js etc.

2. If able, add this capability to the devops pipeline / release output of this project going forward so others can benefit from a plain-JS release without WASM and without the need for dynamic JavaScript execution. This ultimately supports secure computing environments. Success here would be not having to do this step manually again, with the single Empscripem noplugins vanilla non-WASM JS file being available in the release dir and ready for people to use without building the library.


Thank you!
